servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleThreadedExecutor.java

```diff
@@ -0,0 +1,143 @@
+    private final WorkerThread thread;
+
+    SingleThreadedExecutor() {
+        this(DEFAULT_NAME_PREFIX);
+    }
+
+    SingleThreadedExecutor(String namePrefix) {
+        this(namePrefix, true, true);
+    }
+
+    SingleThreadedExecutor(String namePrefix, boolean nonblocking, boolean interruptOnClose) {
+        String name = namePrefix + "-" + INSTANCE_COUNT.incrementAndGet();
+        this.nonblocking = nonblocking;
+        this.interruptOnClose = interruptOnClose;
+        thread = new WorkerThread(this, name);
+        thread.start();
+    }
+
+    @Override
+    public void execute(final Runnable command) {
+        if (Thread.currentThread() == thread) {
+            try {
+                command.run();
+            } catch (Throwable all) {
+                LOGGER.warn("Uncaught throwable from command " + command, all);
+            }
+        } else {
+            try {
+                if (nonblocking) {
+                    if (!thread.queue.offer(command, 100, TimeUnit.MICROSECONDS)) {
+                        throw new RejectedExecutionException("Executor is busy");
+                    }
+                 } else {
+                    thread.queue.put(command);
+                }
+            } catch (InterruptedException woken) {
+                // Interrupt status is not cleared
+                throw new RejectedExecutionException("Enqueue was interrupted", woken);
+            }
+        }
+    }
+
+    @Override
+    public void close() {
+        if (thread.isAlive() &&
+                WorkerThread.stateUpdater.compareAndSet(thread, WorkerThread.OPEN, WorkerThread.CLOSED) &&
+                interruptOnClose &&
+                Thread.currentThread() != thread) {
+            thread.interrupt();
+        }
+    }
+
+    private static class WorkerThread extends AsyncContextHolderThread {
+        static final AtomicIntegerFieldUpdater<WorkerThread> stateUpdater =
+                newUpdater(WorkerThread.class, "state");
+        static final int OPEN = 0;
+        static final int CLOSED = 1;
+        static final int EXITED = 2;
+
+        final SynchronousQueue<Runnable> queue = new SynchronousQueue<>();
+        /**
+         * If the executor is discarded and GCed this reference will be cleared and the worker thread can exit.
+         */
+        private final WeakReference<SingleThreadedExecutor> owner;
+        volatile int state;
+
+        WorkerThread(SingleThreadedExecutor owner, String name) {
+            super(name);
+            this.owner = new WeakReference<>(owner);
+
+            if (!isDaemon()) {
+                setDaemon(true);
+            }
+
+            if (Thread.NORM_PRIORITY != getPriority()) {
+                setPriority(Thread.NORM_PRIORITY);
+            }
+        }
+
+        public void run() {
+            while (OPEN == state && null != owner.get()) {
+                try {
+                    Runnable command = queue.poll(1L, TimeUnit.MINUTES);
```

Comment:
What is the rationale behind polling? Would it not make sense to just call `queue.take()` and block until a task is ready or the thread is interrupted?